Source/Core/DolphinTool/ExtractCommand.cpp

```diff
@@ -0,0 +1,313 @@
+    }
+  }
+  output_file.close();
+}
+
+int Extract(const std::vector<std::string>& args)
+{
+  optparse::OptionParser parser;
+
+  parser.usage("usage: extract [options]...");
+
+  parser.add_option("-i", "--input")
+      .type("string")
+      .action("store")
+      .help("Path to disc image FILE. *")
+      .metavar("FILE");
+  parser.add_option("-o", "--output")
+      .type("string")
+      .action("store")
+      .help("Path to the destination FOLDER. *")
+      .metavar("FOLDER");
+  parser.add_option("-p", "--partition")
+      .type("string")
+      .action("store")
+      .help("Which specific partition you want to extract.");
+  parser.add_option("-s", "--single")
+      .type("string")
+      .action("store")
+      .help("Which specific file/directory you want to extract.");
+  parser.add_option("-l", "--list")
+      .type("string")
+      .action("store")
+      .help("List all files in volume/partition and print it out to a text file (ls.txt).");
+  parser.add_option("-q", "--quiet")
+      .action("store_true")
+      .help("Mute all messages except for errors.");
+  parser.add_option("-g", "--gameonly")
+      .action("store_true")
+      .help("Only extracts the DATA partition.");
+
+  const optparse::Values& options = parser.parse_args(args);
+
+  const bool mute = options.is_set_by_user("mute");
+  const bool gameonly = options.is_set_by_user("gameonly");
+
+  if (!options.is_set("input"))
+  {
+    fmt::println(std::cerr, "Error: No input image set");
+    return EXIT_FAILURE;
+  }
+  const std::string& input_file_path = options["input"];
+
+  const std::string& output_folder_path = options["output"];
+
+  if (!options.is_set("output") && !options.is_set("list"))
+  {
+    fmt::println(std::cerr, "Error: No output folder set");
+    return EXIT_FAILURE;
+  }
+
+  const std::string& single_file_path = options["single"];
+  const std::string& list_path = options["list"];
+  std::string& specific_partition = const_cast<std::string&>(options["partition"]);
+
+  if (options.is_set("output"))
+    std::filesystem::create_directories(output_folder_path);
+
+  if (gameonly)
+    specific_partition = std::string("data");
+
+  if (const std::unique_ptr<DiscIO::BlobReader> blob_reader =
+          DiscIO::CreateBlobReader(input_file_path);
+      !blob_reader)
+  {
+    fmt::println(std::cerr, "Error: Unable to open disc image");
+    return EXIT_FAILURE;
+  }
+
+  const std::unique_ptr<DiscIO::Volume> disc_volume = DiscIO::CreateVolume(input_file_path);
+
+  if (!disc_volume)
+  {
+    fmt::println(std::cerr, "Error: Unable to open volume");
+    return EXIT_FAILURE;
+  }
+
+  if (!VolumeSupported(*disc_volume))
+    return EXIT_FAILURE;
+
+  if (options.is_set_by_user("list"))
+  {
+    ListVolume(*disc_volume, list_path, specific_partition);
+    return EXIT_SUCCESS;
+  }
+
+  bool extracted_one = false;
+
+  if (disc_volume->GetPartitions().empty())
```

Comment:
Also you have two blocks of code here (one for the PARTITION_NONE case and one partition loop) that are practically identical, so please just extract them into a common function.